### DISCUSSION TITLE
Configure Different Template Engine for UI

### DISCUSSION THREAD

**Discussion by Vici37** [martenframework/marten#256] [Ideas]
Hello! Thanks for making the marten framework :) I'm trying out porting one of my personal projects to something less, uh, home grown, and have found marten most closely matches my own preferences for web frameworks. The only gotcha for me is the marten template language, for a couple of reasons:

* Shares syntax with crystal macro language (a footgun for myself, who tends to use the macro language a lot)
* Isn't compile time checked (allowing me to refer to context variables that don't exist)

It would be great if other templating languages or engines could be configured (maybe through something like [kilt](https://github.com/jeromegn/kilt)? Hasn't been updated in awhile, so maybe more of a starting place than a solution) that would give the compile time checks at least, as that's been a wonderful way to bash bugs in my own projects. Even if this idea was accepted, I wouldn't consider this a quick turnaround or anything like that. Just want to throw out the idea.

Thanks again for the hard work in making a great framework!

**Comment by ellmetha** [comment]
Hey! 👋 

While the framework does not explicitly allow the use of other templating solutions, it is perfectly possible to use a solution like Kilt with your language of choice with Marten (for example, you can easily return HTML content generated using another templating language in handlers if necessary).

That said, Marten encourages using its templating engine for all server-side rendering use cases and will never recommend using something else officially. The fact that templates are interpreted at runtime is one of the fundamental design decisions of the framework. This means that errors in templates can't be caught until they are parsed (which happens while rendering them), but the upside obviously is that since templates are not processed at compilation time, you don't need to recompile your whole application in order to update these and this improves the feedback loop in development (which becomes especially relevant with templates-heavy projects).

**Reply by Vici37** [reply]
Aha! Well, this is embarrassing; after reading a considerable amount of the documentation, I completely forgot about the [respond](https://martenframework.com/docs/handlers-and-http/introduction#respond) helper method. That completely solves my own desires at least.

Yup, that reasoning makes complete sense, and I wouldn't want other's to have to deal with compile time when iterating on template heavy development. My own development style focuses much more on the backend and I'm already very comfortable with that type of feedback loop (which also means my templates tend to be on the light side) and prefer those types of errors to be caught at compile time if I can get it.

Thanks again! :D
